scripts/pre_commit_linter.py

```diff
@@ -1061,6 +1071,65 @@ def _check_docstrings(all_files):
                         print '%s --> Line %s: %s' % (
                             filename, line_num + 1, multiline_docstring_message)
 
+    # Check that the args in the docstring are listed in the same
+    # order as they appear in the function definition.
+    def _check_docstrings_arg_order(filename):
```

Comment:
Here and in lines 1083 and 1086, I think it would be better not to define functions within functions. It makes things a bit harder to read, as well as test. Could we define them at the top-level instead, or alternatively make a DocstringChecker manager class to group them all together and define private methods within it?

I also think that, conceptually, it would be simpler to combine the logic here with the stuff above (that checks the formatting of docstrings). I.e. let's find the functions in the file, extract the arg list and docstrings from each function, decompose the docstring into its subcomponents, and validate each part of the docstring. That way, you only need to parse through the file once, and we'd be using an AST-based approach throughout, which makes things more consistent. What do you think?

Reply:
>  How iares the pre-commit linter logic and pylint logic related? If I make changes to one, do I need to make the corresponding change to the other? Is there a way for them to use a single source of truth?

That's a great question -- I'm not 100% sure. @apb7 could you please comment on this?

Reply:
Good question, @lokijuhy! Let me explain you what each of the script is used for:
1. `scripts/pre_commit_linter.py`: This is the main script which houses the logic for all the checks. All new checks are implemented here. One of the checks we have here is the pylint check where we run pylint on all Python files.
2. `scripts/pylint_extensions.py`: As the name suggests, this script contains all custom pylint extensions. These have been written specifically for Oppia and integrated with the pylint check in `scripts/pre_commit_script.py`.
3. `scripts/docstrings_checker.py`: These contain utility functions for docstring checking, customized for the docstring style followed at Oppia. They are used by the `DocstringParameterChecker` in `scripts/pylint_extensions.py`.

You do not need to make any changes to `scripts/pylint_extensions.py` if you're implementing a new check/class in `scripts/pre_commit_linter.py` and vice versa. The former are custom checks/extensions for Pylint and latter is a general script for all checks.
Hope this clears your doubt! If you need any further help, feel free to ping me. Thanks!